parser_hevc.c

```diff
@@ -0,0 +1,387 @@
+  m_PesBufferInitialSize      = 240000;
+
+  m_IsVideo = true;
+  Reset();
+}
+
+cParserHEVC::~cParserHEVC()
+{
+}
+
+void cParserHEVC::Parse(sStreamPacket *pkt, sStreamPacket *pkt_side_data)
+{
+  if (m_PesBufferPtr < 10) // 2*startcode + header + trail bits
+    return;
+
+  int p = m_PesParserPtr;
+  uint32_t startcode = m_StartCode;
+  bool frameComplete = false;
+
+  while ((m_PesBufferPtr - p) > 3)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
+    {
+      if (Parse_HEVC(p, frameComplete) < 0)
+        break;
+    }
+    startcode = startcode << 8 | m_PesBuffer[p++];
+  }
+  m_PesParserPtr = p;
+  m_StartCode = startcode;
+
+  if (frameComplete)
+  {
+    if (!m_NeedSPS && m_FrameValid)
+    {
+      double PAR = (double)m_PixelAspect.num/(double)m_PixelAspect.den;
+      double DAR = (PAR * m_Width) / m_Height;
+      DEBUGLOG("HEVC SPS: PAR %i:%i", m_PixelAspect.num, m_PixelAspect.den);
+      DEBUGLOG("HEVC SPS: DAR %.2f", DAR);
+
+      int duration;
+      if (m_curDTS != DVD_NOPTS_VALUE && m_prevDTS != DVD_NOPTS_VALUE && m_curDTS > m_prevDTS)
+        duration = m_curDTS - m_prevDTS;
+      else
+        duration = m_Stream->Rescale(20000, 90000, DVD_TIME_BASE);
+
+      if (m_FpsScale == 0)
+        m_FpsScale = m_Stream->Rescale(duration, DVD_TIME_BASE, 90000);
+
+      bool streamChange = m_Stream->SetVideoInformation(m_FpsScale, DVD_TIME_BASE, m_Height, m_Width, DAR);
+
+      pkt->id       = m_pID;
+      pkt->size     = m_PesNextFramePtr;
+      pkt->data     = m_PesBuffer;
+      pkt->dts      = m_DTS;
+      pkt->pts      = m_PTS;
+      pkt->duration = duration;
+      pkt->streamChange = streamChange;
+
+    }
+    m_StartCode = 0xffffffff;
+    m_PesParserPtr = 0;
+    m_FoundFrame = false;
+    m_FrameValid = true;
+  }
+}
+
+void cParserHEVC::Reset()
+{
+  cParser::Reset();
+  m_StartCode = 0xffffffff;
+  m_NeedSPS = true;
+  m_NeedPPS = true;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_rbspBufferSize = 0;
+  if (m_rbspBuffer)
+  {
+    free(m_rbspBuffer);
+    m_rbspBuffer = NULL;
+  }
+}
+
+int cParserHEVC::Parse_HEVC(int buf_ptr, bool &complete)
+{
+  unsigned int len = m_PesBufferPtr - buf_ptr;
+  uint8_t *buf = m_PesBuffer + buf_ptr;
+  unsigned int p=0, NumBytesInNalUnit=0;
+
+  // find length of nal (including nal_unit_header)
+  uint32_t startcode = 0xffffffff;
+  while ((p+3) < len)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
+    {
+      NumBytesInNalUnit = p-3;  
+      break;
+    }
+    startcode = startcode << 8 | buf[p++];
+  }
+  if (!NumBytesInNalUnit)
+    return -1;
+
+  // create buffer for rbsp
+  if (NumBytesInNalUnit > m_rbspBufferSize)
+  {
+    if (NumBytesInNalUnit >= 500000)
+    {
+      ERRORLOG("cParserHEVC::Parse_HEVC - max buffer size reached, pid: %d", m_pID);
+      Reset();
+      return 0;
+    }
+    m_rbspBufferSize = NumBytesInNalUnit;
+    uint8_t *new_buffer = (uint8_t*)realloc(m_rbspBuffer, m_rbspBufferSize);
```

Comment:
No, by no means I wanted to alter the stream that is sent to Kodi, hence the second buffer. 
I agree with your your packet accumulation above, not very good. 
But in order to be able to parse the NAL units, the emulation_prevention_three_byte sequences must be removed first. For the smaller units, this is no problem, as they fit into the TS packets. The VLC packets can be larger though...
However, if I understand the spec correctly (Section 7.4.2.1), the RBSP must be converted to SODB first, otherwise more_rbsp_data() will not work.
I am not sure if the data that is decoded currently is sufficient. Maybe, for getting the required data, a call to more_rbsp_data() is never needed, so one might remove the emulation_prevention_three_byte on the fly. No copying anymore, very nice.
But this would require a different version of the cBitstream bs functions that are aware of the emulation_prevention_three_byte and skips them. I didn't dare to change cBitstream at this time.